Docs: New Frontend System - Plugin architecture (#22325)
backstage/backstage: docs/frontend-system/architecture/04-plugins.md

## Introduction

In addition to the existing [`plugins`](../../plugins/index.md) documentation let's take a look at what changes for plugins in the new frontend system. If you already created a plugin yourself you will recognise a lot of similarity in the new architecture with the existing one.

> Backstage is a single-page application composed of a set of plugins. Each of this plugins should solve exactly one responsibility & follow our suggested [plugin package structure & naming](../../architecture-decisions/adr011-plugin-package-structure.md).

Reviewer: I think "solve exactly one responsibility" might cause more questions than it answers tbh. Might be worth be worth switch it to something like "Each of these plugins have have solutions for their own set of responsibilities, and should follow ..."

> For frontend plugins each plugin should only export a single plugin instance. The new frontend system can detect plugins, if they are exported as `default`, from the plugin package. This is also possible during runtime allowing you to add plugins to your Backstage instance without requiring a restart.

Reviewer: ```suggestion
Frontend plugins should only export a single plugin instance as its `default` export. This simplifies installation of the plugin, as the framework is able to discover default exports automatically.
```

Couple of tweaks, although it might need a bit more work since it got a bit short, maybe merge without the previous paragraph etc.

This is all related to frontend plugins, so no need to specify it as explicitly imo. Similarly, this is all also talking about the new frontend system. On that note it's best to keep these docs as standalone as possible so that we don't need to rewrite them again in the future. So rather than explaining things in terms of the existing system, try to go for a standalone explanation. We do still have the migration sections where we can explain things in terms of the old system.

Lastly I dropped the last sentence, that's referring to dynamic plugin installation right? We don't support that yet so best not to imply that we do. Although we could bring it back as a future promise though, although we'll definitely forget to update that xP

## Creating a Plugin

A plugin is can easily be created, it only requires a plugin `id`:

```ts title="plugins/tech-radar/src/index.ts"
export const techRadarPlugin = createPlugin({
  id: 'tech-radar',
});
```

The created plugin does not yet do anything, it will need some plugin options to be useful:

- Extend it's functionality, through pages, navigation items or entity cards, using `extensions`
- You can make it reachable through `routes`
- Link to other plugins in Backstage using `externalRoutes`
- Pass `featureFlags` to it

### Plugin ID

> You will decide on the `id` when you [create a plugin](../../). The plugin `id` should follow the respective [naming pattern](./08-naming-patterns.md#plugins) of the new frontend system. To give you an example the GraphiQL plugin has the `id` `tech-radar` & as it follows the naming convention we know that the `techRadarPlugin` is the symbol for the default package export containing the plugin.

Reviewer 2: ```suggestion
You will decide on the ID when you [create a plugin](../../). The plugin ID should follow the respective [naming pattern](./08-naming-patterns.md#plugins) of the new frontend system. To give you an example the Tech Radar plugin has the ID `tech-radar`, and as it follows the naming convention we know that the `techRadarPlugin` is the symbol for the default package export containing the plugin.
```

i don't know if it's just me, but the ampersands stand out :)

### Plugin Extensions

> So let's make our plugin a bit more useful! Imagine we want to have a TechRadar plugin that displays the recommended technologies at our organisation, [just like in the Backstage demo instance](https://demo.backstage.io/tech-radar). We want the plugin to be displayed on a page & have a navigation item directing to it. This can be achieved by creating a page extension & a navigation item extension. When adding those to the plugin they will be imported with the plugin package & can be discovered.

Reviewer: I think this and many of the paragraphs below read perhaps a little bit too much like a guide rather than a description of the architecture. This section of the docs is more geared towards explaining how things work and are wired together, rather than a step-by-step guide to building a plugin.

```ts title="plugins/tech-radar/src/index.ts"
export const techRadarPlugin = createPlugin({
  id: 'tech-radar',
  extensions: [techRadarPage, techRadarNavItem],
});
```

### Plugin Routes

Now that we have a page extension in our plugin the route of the page is detected & the page is available. Thought to make the route easy accessible through the plugin you are required to additionally add those routes on the plugin level. If we have a `techRadarRootRef`, that is the route reference for the `techRadarPage`, we would add it to `createPlugin` like the following:

```ts title="plugins/tech-radar/src/index.ts"
export const techRadarPlugin = createPlugin({
  id: 'tech-radar',
  extensions: [techRadarPage, techRadarNavItem],
  routes: {
    root: techRadarRootRef,
  },
});
```

> ### Plugin External Routes

Reviewer: External routes are already explained in the routing architecture docs. I think it's perhaps best to collapse this and the previous section into a single plugin routes docs that covers both, but quickly links over to the routing docs for more details?

Author: not sure what you mean by this. should I create a seperate `0x-plugin-routes.md` file in the `architecture` folder?

Reviewer: There is one already, it's all documented here: https://backstage.io/docs/frontend-system/architecture/routes/

I still think this section is useful, but that it would be best to treat it just as a very quick introduction to routes and then quickly linking over to the routing architecture docs instead.

});
```

```ts title="plugins/tech-radar/src/index.ts"
export const techRadarPlugin = createPlugin({
  id: 'tech-radar',
  extensions: [techRadarPage, techRadarNavItem],
  routes: {
    root: techRadarRootRef,
  },
  externalRoutes: [
    {
      external: externalRouteRef,
    },
  ],
});
```

A route ref to an external route can be than used just like an internal route ref.

```tsx title="plugins/tech-radar/src/components/SomePage.tsx"
...
const SomePage = createPageExtension({
  ...
  loader: async () => {
    const Component = () => {
      const externalLink = useRouteRef(externalRouteRef);

      ...
      return (
        ...
        <Link to={externalLink()}>External Page</Link>
        ...
      );
    }
    return <Component />;
  }
...
```

#### External Routes Configuration

Something special with the external routes is, that they will be discovered and can be configured directly through the `app-config.yaml` outside of the specific plugin context. Imagine that the TechRadar has an external route that links to a documentation on the TechRadar. You could configure the external route like the following:

```yaml title="app-config.yaml"
app:
  routes:
    bindings:

>       plugin.techRadar.externalRoutes.external: plugin.other.routes.docOnTechRadar

Reviewer 2: ```suggestion
     techRadar.external: other.docOnTechRadar
```

the format changed

```

### Plugin Feature Flags

With the `featureFlags` array we can pass feature flags into the plugin. Imagine you are planning to update your TechRadar, but you only want to allow people with the `show-future-tech` flag to see the current draft to give feedback. Your plugin could expect the flag like the following:

```ts title="plugins/tech-radar/src/index.ts"
export const techRadarPlugin = createPlugin({
  id: 'tech-radar',
  extensions: [techRadarPage, techRadarNavItem],

>   featureFlags: [{ showFutureTech: 'show-future-tech' }],

Reviewer 2: ```suggestion
  featureFlags: [{ name: 'show-future-tech' }],
```

at least in the new system
